runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -147,12 +152,22 @@ public SparkPipelineResult run(final Pipeline pipeline) {
 
     MetricsEnvironment.setMetricsSupported(true);
 
+    // visit the pipeline to determine the translation mode
     detectTranslationMode(pipeline);
 
+    // create the cache candidates map
+    Map<PCollection, Long> cacheCandidates = new HashMap<>();
```

Comment:
What if this was removed entirely ? I think you can use the `EvaluationContext`'s cachedCandidates directly because now the cache visitor has access to it.